data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepositoryDDB.java

```diff
@@ -76,4 +83,81 @@ public Environment getEnvironment(
   }
+
+  @Override
+  public Environment createEnvironmentTargetVersion(final Environment environment)
+      throws StorageException, EnvironmentExistsException {
+
+    final EnvironmentTargetVersionDDBRecord environmentTargetVersionDDBRecord =
+        environmentMapper.toEnvironmentTargetVersionDDBRecord(environment);
+
+    try {
+      dynamoDBMapper.save(environmentTargetVersionDDBRecord);
+    } catch (final AmazonServiceException e) {
+      throw new StorageException(
+          String.format(
+              "Could not save record with environment id %s", environment.getEnvironmentId()),
+          e);
+    }
+    return environmentMapper.toEnvironment(environmentTargetVersionDDBRecord);
+  }
+
+  @Override
+  public Environment getEnvironmentTargetVersion(String environmentId)
+      throws StorageException, EnvironmentNotFoundException {
+
+    final EnvironmentTargetVersionDDBRecord environmentTargetVersionDDBRecord;
+    try {
+      environmentTargetVersionDDBRecord =
+          dynamoDBMapper.load(EnvironmentTargetVersionDDBRecord.withHashKey(environmentId));
+    } catch (final AmazonServiceException e) {
+      throw new StorageException(
+          String.format("Could not load record with environment id %s", environmentId), e);
+    }
+
+    if (environmentTargetVersionDDBRecord == null) {
+      throw new EnvironmentNotFoundException(
+          String.format("Could not find environment with id %s", environmentId));
+    }
+
+    return environmentMapper.toEnvironment(environmentTargetVersionDDBRecord);
+  }
+
+  @Override
+  public List<String> listClusters() throws StorageException {
+    try {
+      List<EnvironmentTargetVersionDDBRecord> scanResult =
+          dynamoDBMapper.scan(
+              EnvironmentTargetVersionDDBRecord.class, new DynamoDBScanExpression());
+      //TODO: handle pagination. paginatedlist returns more elements on demand does stream accomplish that?
+      return scanResult.stream().map(e -> e.getCluster()).distinct().collect(Collectors.toList());
+    } catch (final AmazonServiceException e) {
+      throw new StorageException("Could not scan environment target versions");
+    }
+  }
+
+  @Override
+  public List<String> listEnvironmentIdsByCluster(@NonNull final String cluster)
+      throws StorageException {
+    DynamoDBQueryExpression<EnvironmentTargetVersionDDBRecord> queryExpression =
+        new DynamoDBQueryExpression<EnvironmentTargetVersionDDBRecord>()
+            .withIndexName(EnvironmentTargetVersionDDBRecord.ENVIRONMENT_CLUSTER_GSI_NAME)
+            .withConsistentRead(false)
+            .withHashKeyValues(EnvironmentTargetVersionDDBRecord.withGSIHashKey(cluster));
+
+    try {
+      //TODO: handle pagination. paginatedlist returns more elements on demand does stream accomplish that?
+      List<EnvironmentTargetVersionDDBRecord> queryResult =
+          dynamoDBMapper.query(EnvironmentTargetVersionDDBRecord.class, queryExpression);
+
+      if (queryResult.isEmpty()) {
```

Comment:
Why the special edge case? AFAICT, the mapping on line 157 will have the same effect if the source list is empty.

Reply:
oh interesting, the test failed previously, I must've done something else wrong